bigrig/config.py

```diff
@@ -0,0 +1,202 @@
+                    "type": "string",
+                    "description": "Local FS folder or url to pypi simple compatible repo",
+                },
+                "credentialsPath": {
+                    "type": "string",
+                    "description": "Path to YAML file containing CREDENTIALS_SCHEMA",
+                },
+            },
+            "required": ["location"],
+            "additionalProperties": False,
+        },
+        "source": {
+            "type": "object",
+            "description": "where we put sources fetched from origin for internal keeping. Need only one of those",
+            "properties": {
+                "location": {
+                    "type": "string",
+                    "description": "url to pypi simple compatible repo",
+                }
+            },
+            "required": ["location"],
+            "additionalProperties": False,
+        },
+        "target": {
+            "type": "object",
+            "properties": {
+                "variables": {
+                    "type": "object",
+                    "description": "Variables to be forwarded to the dockerfile template and used inside",
+                },
+                "location": {
+                    "type": "string",
+                    "description": "url to pypi simple compatible repo",
+                },
+            },
+            "required": ["variables", "location"],
+            "additionalProperties": False,
+        },
+    },
+    "properties": {
+        "origin": {"$ref": "#/definitions/origin"},
+        "source": {"$ref": "#/definitions/source"},
+        "targets": {
+            "type": "object",
+            "description": "where we put the built things. As many as we have arches + pythonversions ( or maybe just arches? You can have several pythonversions coexist in one repo)",
+            "patternProperties": {"^[a-z0-9_]+$": {"$ref": "#/definitions/target"}},
+        },
+    },
+}
+
+
+@dataclass
+class Origin:
+    location: str
+    credentials: t.Optional["OriginCredentials"]
+
+    @classmethod
+    def from_dict(cls, blob: t.Dict) -> "Origin":
+        return cls(
+            location=blob["location"],
+            credentials=OriginCredentials.from_path(path=blob.get("credentialsPath")),
+        )
+
+
+@dataclass
+class OriginCredentials:
+    username: str
+    password: str
+
+    @classmethod
+    def from_path(cls, path: t.Optional[str]) -> t.Optional["OriginCredentials"]:
+        if not path:
+            return None
+
+        try:
+            with open(path, "rt") as fid:
+                credentials = yaml.safe_load(fid)
+        except yaml.YAMLError:
+            # Raise from None to avoid accidental leaks
+            raise yaml.YAMLError(
+                f"Unable to load credentials from `{path}`, it appears to be a non-yaml file"
+            ) from None
+        except OSError as exc:
+            raise OSError(f"Unable to load credentials from `{path}`") from exc
```

Comment:
Nitpick: is there a lot of point of catching and reraising the same error with lightly different message?

Reply:
As discussed via slack, I tend to catch these kind of exceptions to provide more user-friendly messages, while I agree that this one in particular is just the same thing, I'd like to leave it as is for consistency with the rest of the module.